server/handler/query.go

```diff
@@ -84,17 +93,17 @@ func Query(db service.SQLDB) RequestProcessFunc {
 		// case. We check if the context was cancelled with Err() instead of Done()
 		select {
 		case <-r.Context().Done():
+			if r.Context().Err() != nil {
```

Comment:
Is it safe to move this here from the previous place, after the `select{}`?

As the comment above says, if the context is cancelled this select may get into the `case err` branch. In that case, the `KILL` will not be executed.

Reply:
I explored what's happening here and I also tried to simulate it [here](https://play.golang.org/p/aj8afgoaiCP).

There are 3 possible outcomes:
1. The query runs and returns no error,
2. The query runs and returns an error,
3. The query runs and is cancelled by the user.

In the first case the `select` unblocks as it receives `nil` in channel `c`.
In the second case the `select` unblocks as it receives a non-`nil` value in channel `c`.
In the third case the `select` unblocks as the cancelled context becomes `Done`, and in this case `r.Context().Err() != nil`. Additionally from the snippet that I wrote it the `err` channel is even never written in this case (unless the outcome of the writing is not deterministic, could that be the case?).

Reply:
Addressed [here](https://github.com/src-d/gitbase-web/pull/305/commits/1ed3bc62e2e9474f72fedae0e40277fccb1f23ae)

Reply:
For scenario 3, what can happen is that the `conn.QueryContext` will stop when the context is cancelled and return an error. Then in this select, you will enter any of the 2 possible `case`, and it's not deterministic. From the [go docs](https://tour.golang.org/concurrency/5): 

> A select blocks until one of its cases can run, then it executes that case. It chooses one at random if multiple are ready.

You can test this in your branch, run this:
```
watch -n 0.1 -e "GOCACHE=off go test -v github.com/src-d/gitbase-web/server/handler -run TestQuerySuite/TestQueryAbort"
```

And sooner or later you will get a failure:
```
=== RUN   TestQuerySuite
=== RUN   TestQuerySuite/TestQueryAbort
--- FAIL: TestQuerySuite (0.20s)
    --- FAIL: TestQuerySuite/TestQueryAbort (0.20s)
        require.go:157:
                Error Trace:    query_test.go:244
                                                        server.go:1964
                                                        request_logger.go:39
                                                        server.go:1964
                                                        query_test.go:249
                                                        asm_amd64.s:1333
                Error:          Not equal:
                                expected: *errors.errorString(&errors.errorString{s:"context canceled"})
                                actual  : serializer.httpError(serializer.httpError{Status:400, Title:"canceling query due to user request", Details:"", MySQLCode:0x0})
                Test:           TestQuerySuite/TestQueryAbort
        common_unit_test.go:45:
                Error Trace:    common_unit_test.go:45
                                                        suite.go:98
                                                        suite.go:103
                Error:          there were unfulfilled expectations:
                Test:           TestQuerySuite/TestQueryAbort
                Messages:       there is a remaining expectation which was not matched: ExpectedExec => expecting Exec or ExecContext which:
                                  - matches sql: 'KILL 1288'
                                  - is without arguments
FAIL
FAIL    github.com/src-d/gitbase-web/server/handler     0.229s
```

Note: the `Not equal` check is correct, but error message is wrong. It looks like we have the arguments for `require.Equal(context.Canceled, err)` reversed in the test, it should be `require.Equal(err, context.Canceled).

Reply:
Good think you noticed! I haven't thought that the context could be cancelled just **after** the channel has been written, but **before** entering the select! I'm gonna fix it and thanks for explanation @carlosms!

Reply:
Addressed [here](https://github.com/src-d/gitbase-web/pull/305/commits/7e62f4436c93423abf5193a9a5611b4b54cf5bc3).